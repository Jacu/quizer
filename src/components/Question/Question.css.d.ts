export const Question: string;
export const Answers: string;
export const Score: string;
