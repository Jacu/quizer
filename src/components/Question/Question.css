.Question {
    position: relative;
    height: 40vh;
    background: white;
    border-radius: 50px;
    box-shadow: 2px 2px 10px 2px rgba(0,0,0,0.25);
    flex: 6;
    display: flex;
    justify-content: center;
    flex-direction: column;
    overflow: hidden;
}

h1 {
    padding: 1.5rem;
}

h3 {
    font-size: 1rem;
    padding: 0.5rem;
    padding-bottom: 2rem;   
    font-style: italic; 
}
.Answers {
    display: flex;
}