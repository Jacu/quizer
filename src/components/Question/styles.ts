import styled from "styled-components";

export const QuestionComponent = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    flex-direction: column;
    overflow: hidden;
`;

export const QuestionHeader = styled.div`
   display: flex;
`;

export const Question = styled.h3`
    display: flex;
    margin: 15px 0;
`;

export const CategoryLabel = styled.p` 
    font-style: italic;
    display: flex;
`;

export const AnswersContainer = styled.div`
    display: flex;
    flex-direction: column;
`;