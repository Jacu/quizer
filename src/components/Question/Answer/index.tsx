import React from 'react';
import * as styled from './styles';

interface IAnswerProps {
    answer: string,
    isSelected: boolean,
    isCorrect: boolean,
    reveal: boolean,
    onClick: () => void,
}

const Answer: React.FC<IAnswerProps> = ({isSelected, onClick, answer, isCorrect, reveal}) => 
    <styled.Answer 
        isSelected={isSelected}
        isCorrect={isCorrect}
        showAnswer={reveal}
        onClick={!reveal ? onClick : undefined} >
        <styled.Checkbox showAnswer={reveal} isCorrect={isCorrect} isSelected={isSelected} />
        {answer} 
    </styled.Answer>

export default Answer;