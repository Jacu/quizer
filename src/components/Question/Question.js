import React, { Component } from 'react';
import styles from './Question.css';
import { connect } from 'react-redux';
import Answer from '../Answer/Answer';
import * as actions from '../../store/actions/quiz';
// import { Redirect } from 'react-router-dom';

class Question extends Component {
    render() {
        let allAnswers = [...this.props.question[this.props.match.params.id - 1].incorrect_answers];
        allAnswers.splice(this.props.answers[this.props.match.params.id - 1] - 1, 0, this.props.question[this.props.match.params.id - 1].correct_answer);

        const redirect = (
            Number(this.props.match.params.id) <= 0 || Number(this.props.match.params.id) > this.props.totalQuestions ? this.props.history.goBack() : null
        )
        return (
            <div className={styles.Question}>
                {redirect}
                <h1>Question {this.props.match.params.id}/{this.props.totalQuestions}</h1>
                <h3>Category: {this.props.question[this.props.match.params.id - 1].category}</h3>
                <p>{this.props.question[this.props.match.params.id - 1].question}</p>
                <div className={styles.Answers}>
                    {
                        allAnswers.map((answer, i) => (
                            <Answer key={i} answer={answer} selected={this.props.questionsPicked[this.props.match.params.id - 1] === i+1} onClick={() => this.props.pickAnswer(this.props.match.params.id - 1,i+1)} />
                        ))
                    }
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        question: [...state.questions],
        totalQuestions: state.selectedSetting.amount,
        answers: [...state.questionsAnswers],
        questionsPicked: [...state.questionsPicked]
    }
}

const mapDispatchToProps = dispatch => {
    return {
        pickAnswer: (index, newPick) => dispatch(actions.pickAnswer(index, newPick))
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(Question);