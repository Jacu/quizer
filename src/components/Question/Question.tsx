import React from 'react';
import styles from './Question.css';
import { connect } from 'react-redux';
import Answer from '../Answer/Answer';
import * as actions from '../../store/actions/quiz';
import unescape from '@favware/unescape';

interface Question {
    totalQuestions, 
    questions, 
    questionsPicked, 
    correctAnswers, 
    ended
}

const Question = (props) =>  {
    const {totalQuestions, questions, questionsPicked, correctAnswers, ended } = props;
    const questionId = props.match.params.id - 1;
    const question = questions[questionId];
    const correctAnswerId = correctAnswers[questionId];
    const selectedAnswerId = questionsPicked[questionId];
    const allAnswers = [...question.incorrect_answers];
    allAnswers.splice(correctAnswerId, 0, question.correct_answer);

    const redirect = (
        Number(questionId) < 0 || Number(questionId) > totalQuestions - 1 ? props.history.goBack() : null
    )

    const handleAnswerPick = (answerId) => {
        return () => props.pickAnswer(questionId, answerId)
    }

    const score = !ended ? null :
        (<div className={styles.Score}>{correctAnswerId === selectedAnswerId ? 'GREAT!' : 'BAD LUCK'}</div>)

    return (
        <div className={styles.Question}>
            {redirect}
            <h1>Question {questionId + 1}/{totalQuestions}</h1>
            <h3>Category: {question.category}</h3>
            <p>{unescape(question.question)}</p>
            <div className={styles.Answers}>
                {
                    allAnswers.map((answer, i) => (
                        <Answer 
                            key={i}
                            answer={unescape(answer)}
                            selected={selectedAnswerId === i}
                            onClick={handleAnswerPick(i)} 
                            ended={ended}
                            correct={correctAnswerId === i}
                            />
                    ))
                }
            </div>
            {score}
        </div>
    );
};

const mapStateToProps = ({quiz}) => {
    return {
        questions: [...quiz.questions.all],
        totalQuestions: quiz.questions.amount,
        correctAnswers: [...quiz.answers.correct],
        questionsPicked: [...quiz.answers.picked],
        ended: quiz.finished
    }
}

const mapDispatchToProps = dispatch => {
    return {
        pickAnswer: (index, newPick) => dispatch(actions.pickAnswer(index, newPick))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Question);