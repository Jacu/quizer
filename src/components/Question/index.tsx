import React from 'react';
import * as styled from './styles';
import Answer from './Answer';

interface QuestionProps {
    question: string,
    category: string,
    answers: string[],
    correctId: number,
    reveal: boolean,
    selectedId?: number,
    onAnswerPicked: (answerId: number) => void,
}

const Question: React.FunctionComponent<QuestionProps> = ({ question, category, answers, correctId, reveal, onAnswerPicked, selectedId }) =>
    (
        <styled.QuestionComponent>
            <styled.CategoryLabel>{`Category: ${category}`}</styled.CategoryLabel>
            <styled.Question>{question}</styled.Question>
            <styled.AnswersContainer>
                {answers.map((answer, i) => (
                    <Answer
                        key={i}
                        answer={answer}
                        selected={(selectedId == null && reveal) || selectedId === i || (reveal && correctId === i)}
                        correct={correctId === i}
                        reveal={reveal}
                        onClick={() => onAnswerPicked(i)} />
                ))}
            </styled.AnswersContainer>
        </styled.QuestionComponent>
    );

export default Question;