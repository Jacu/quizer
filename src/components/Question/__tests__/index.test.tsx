import React from 'react';
import Question from '../';
import { render, fireEvent } from '@testing-library/react';

const onAnswerPicked = jest.fn();
const answers = ['a1','a2'];
const props = {
    question: 'Question',
    category: 'Category',
    answers,
    correctId: 0,
    reveal: false,
    //selectedId: 5,
    onAnswerPicked,
}

describe('SettingPanel', () => {
    beforeEach(() => {
        jest.clearAllMocks();
        jest.resetModules();
    });
    it('should render with default props', () => {
        const { asFragment, getByText } = render(<Question {...props} />);
        fireEvent.click(getByText(answers[0]));
        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onAnswerPicked for each selection', () => {
        const { getByText } = render(<Question {...props} />);
        fireEvent.click(getByText(answers[0]));
        fireEvent.click(getByText(answers[1]));
        expect(onAnswerPicked).toHaveBeenCalledTimes(2);
    });

    describe('should call onAnswerPicked with questionId', () => {
        it('when id is 0 (falsy)', () => {
            const { getByText } = render(<Question {...props} />);
            fireEvent.click(getByText(answers[0]));
            expect(onAnswerPicked).toHaveBeenCalledWith(0);
        });
        it('when id is 1 (truthy)', () => {
            const { getByText } = render(<Question {...props} />);
            fireEvent.click(getByText(answers[1]));
            expect(onAnswerPicked).toHaveBeenCalledWith(1);
        });
    });
});