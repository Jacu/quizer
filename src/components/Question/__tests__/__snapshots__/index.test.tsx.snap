// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingPanel should render with default props 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm fbvjrF"
  >
    <p
      class="sc-AxhCb kcDBNW"
    >
      Category: Category
    </p>
    <h3
      class="sc-AxiKw cWgrDT"
    >
      Question
    </h3>
    <div
      class="sc-AxhUy kRznDN"
    >
      <div
        class="sc-AxgMl hrRWqd"
      >
        <div
          class="sc-AxheI eUhxBE"
        />
        a1
      </div>
      <div
        class="sc-AxgMl hrRWqd"
      >
        <div
          class="sc-AxheI eUhxBE"
        />
        a2
      </div>
    </div>
  </div>
</DocumentFragment>
`;
